File: src/retriever.ts
/**
 * Clarivo Local Vector Retriever (Track PS04)
 * TypeScript implementation matching src/retriever.py
 */

import fs from 'fs';
import path from 'path';

export interface KBArticle {
  id: string;
  category: string;
  title: string;
  body: string;
  applies_to: string[];
  required_fields: string[];
  keywords: string[];
}

export interface RetrievalResult {
  id: string;
  title: string;
  category: string;
  body: string;
  applies_to: string[];
  required_fields: string[];
  cosine_similarity: number;
  keyword_score: number;
  score: number;
  passes_threshold: boolean;
}

const DATA_DIR = path.resolve(process.cwd(), 'data');
const EMBEDDINGS_FILE = path.join(DATA_DIR, 'kb_embeddings.json');
const ARTICLES_FILE = path.join(DATA_DIR, 'kb_articles.json');
export const SIMILARITY_THRESHOLD = 0.65;

let indexCache: Record<string, any> | null = null;
let articlesCache: Record<string, KBArticle> | null = null;

export function loadKBData(): { index: Record<string, any>; articles: Record<string, KBArticle> } {
  if (!indexCache) {
    if (fs.existsSync(EMBEDDINGS_FILE)) {
      indexCache = JSON.parse(fs.readFileSync(EMBEDDINGS_FILE, 'utf-8'));
    } else {
      indexCache = {};
    }
  }

  if (!articlesCache) {
    if (fs.existsSync(ARTICLES_FILE)) {
      const list: KBArticle[] = JSON.parse(fs.readFileSync(ARTICLES_FILE, 'utf-8'));
      articlesCache = {};
      for (const a of list) {
        articlesCache[a.id] = a;
      }
    } else {
      articlesCache = {};
    }
  }

  return { index: indexCache!, articles: articlesCache! };
}

export function cosineSimilarity(a: number[], b: number[]): number {
  let dot = 0;
  let normA = 0;
  let normB = 0;
  const len = Math.min(a.length, b.length);
  for (let i = 0; i < len; i++) {
    dot += a[i] * b[i];
    normA += a[i] * a[i];
    normB += b[i] * b[i];
  }
  const denominator = Math.sqrt(normA) * Math.sqrt(normB);
  return denominator === 0 ? 0 : dot / denominator;
}

export function calculateKeywordScore(query: string, keywords: string[], title: string): number {
  const q = query.toLowerCase();
  const queryTokens = new Set(
    q
      .split(/\s+/)
      .map((w) => w.replace(/[?,.!]/g, ''))
      .filter((w) => w.length > 2)
  );

  let matches = 0.0;
  for (const kw of keywords || []) {
    const kwLower = kw.toLowerCase();
    if (q.includes(kwLower)) {
      matches += 3.0;
    } else {
      const kwTokens = kwLower.split(/\s+/);
      for (const t of kwTokens) {
        if (queryTokens.has(t)) {
          matches += 1.0;
        }
      }
    }
  }

  const titleWords = title.toLowerCase().split(/\s+/);
  for (const w of titleWords) {
    const clean = w.replace(/[?,.!]/g, '');
    if (clean.length > 3 && queryTokens.has(clean)) {
      matches += 1.0;
    }
  }

  if (matches >= 3.0) {
    return Math.min(0.96, 0.78 + matches * 0.03);
  } else if (matches >= 1.5) {
    return Math.min(0.85, 0.72 + matches * 0.04);
  } else if (matches > 0) {
    return Math.min(0.68, 0.4 + matches * 0.1);
  }
  return 0.1;
}

export function retrieveKB(
  queryText: string,
  queryVector?: number[] | null,
  topK = 3,
  categoryFilter?: string
): RetrievalResult[] {
  const { index, articles } = loadKBData();
  const results: RetrievalResult[] = [];

  for (const [id, item] of Object.entries(index)) {
    if (categoryFilter && item.category !== categoryFilter) {
      continue;
    }

    const fullArticle = articles[id] || {
      id,
      category: item.category,
      title: item.title,
      body: '',
      applies_to: [],
      required_fields: [],
      keywords: item.keywords || [],
    };

    const kwScore = calculateKeywordScore(queryText, item.keywords || [], item.title || '');
    let denseSim = 0;
    let combinedScore = kwScore;

    if (queryVector && item.vector && Array.isArray(item.vector)) {
      denseSim = cosineSimilarity(queryVector, item.vector);
      combinedScore = 0.75 * denseSim + 0.25 * kwScore;
    }

    results.push({
      id,
      title: item.title,
      category: item.category,
      body: fullArticle.body,
      applies_to: fullArticle.applies_to,
      required_fields: fullArticle.required_fields,
      cosine_similarity: Number(denseSim.toFixed(4)),
      keyword_score: Number(kwScore.toFixed(4)),
      score: Number(combinedScore.toFixed(4)),
      passes_threshold: combinedScore >= SIMILARITY_THRESHOLD,
    });
  }

  results.sort((a, b) => b.score - a.score);
  return results.slice(0, topK);
}

File: src/retriever.py
#!/usr/bin/env python3
"""
Clarivo Local Vector Retriever (Track PS04)
Performs local similarity search over precomputed gemini-embedding-001 vectors.
Combines dense vector similarity with deterministic keyword validation.
"""

import os
import json
import math
from typing import List, Dict, Any, Tuple, Optional

DATA_DIR = os.path.join(os.path.dirname(os.path.dirname(__file__)), "data")
EMBEDDINGS_FILE = os.path.join(DATA_DIR, "kb_embeddings.json")
ARTICLES_FILE = os.path.join(DATA_DIR, "kb_articles.json")

# Similarity threshold: queries scoring below this are considered uncovered
SIMILARITY_THRESHOLD = 0.65

_INDEX_CACHE: Optional[Dict[str, Any]] = None

def load_kb_index() -> Dict[str, Any]:
    global _INDEX_CACHE
    if _INDEX_CACHE is not None:
        return _INDEX_CACHE

    if not os.path.exists(EMBEDDINGS_FILE):
        raise FileNotFoundError(f"Embedding cache not found at {EMBEDDINGS_FILE}")

    with open(EMBEDDINGS_FILE, "r", encoding="utf-8") as f:
        _INDEX_CACHE = json.load(f)
    return _INDEX_CACHE

def cosine_similarity(vec_a: List[float], vec_b: List[float]) -> float:
    dot_product = sum(a * b for a, b in zip(vec_a, vec_b))
    norm_a = math.sqrt(sum(a * a for a in vec_a))
    norm_b = math.sqrt(sum(b * b for b in vec_b))
    if norm_a == 0 or norm_b == 0:
        return 0.0
    return dot_product / (norm_a * norm_b)

def keyword_score(query: str, article_keywords: List[str], title: str, body: str) -> float:
    query_lower = query.lower()
    query_tokens = set([w.strip("?,.!") for w in query_lower.split() if len(w) > 2])

    matches = 0.0
    for kw in article_keywords:
        kw_lower = kw.lower()
        if kw_lower in query_lower:
            matches += 3.0
        else:
            # Check token overlap
            kw_tokens = set(kw_lower.split())
            overlap = query_tokens.intersection(kw_tokens)
            if overlap:
                matches += 1.0 * len(overlap)

    for word in title.lower().split():
        clean_word = word.strip("?,.!")
        if len(clean_word) > 3 and clean_word in query_tokens:
            matches += 1.0

    # Calibrated score: if there are strong keyword matches (matches >= 2.0), score is in 0.75-0.95 range
    if matches >= 3.0:
        return min(0.96, 0.78 + (matches * 0.03))
    elif matches >= 1.5:
        return min(0.85, 0.72 + (matches * 0.04))
    elif matches > 0:
        return min(0.68, 0.40 + (matches * 0.10))
    return 0.10

def retrieve_kb(
    query_text: str,
    query_vector: Optional[List[float]] = None,
    top_k: int = 3,
    category_filter: Optional[str] = None
) -> List[Dict[str, Any]]:
    """
    Retrieves the top-k relevant KB articles.
    If query_vector is provided, calculates hybrid (dense cosine + lexical).
    Otherwise calculates lexical match.
    """
    index = load_kb_index()
    # Also load full article bodies
    with open(ARTICLES_FILE, "r", encoding="utf-8") as f:
        articles_data = {a["id"]: a for a in json.load(f)}

    results = []
    for art_id, item in index.items():
        if category_filter and item.get("category") != category_filter:
            continue

        full_art = articles_data.get(art_id, {})
        kw_score = keyword_score(
            query_text,
            item.get("keywords", []),
            item.get("title", ""),
            full_art.get("body", "")
        )

        if query_vector and item.get("vector"):
            dense_sim = cosine_similarity(query_vector, item["vector"])
            # Hybrid fusion
            combined_score = (0.75 * dense_sim) + (0.25 * kw_score)
        else:
            dense_sim = 0.0
            combined_score = kw_score

        results.append({
            "id": art_id,
            "title": item.get("title"),
            "category": item.get("category"),
            "body": full_art.get("body", ""),
            "applies_to": full_art.get("applies_to", []),
            "required_fields": full_art.get("required_fields", []),
            "cosine_similarity": round(dense_sim, 4),
            "keyword_score": round(kw_score, 4),
            "score": round(combined_score, 4),
            "passes_threshold": combined_score >= SIMILARITY_THRESHOLD
        })

    results.sort(key=lambda x: x["score"], reverse=True)
    return results[:top_k]

if __name__ == "__main__":
    test_query = "Why is my internet down today? Did I miss a payment?"
    print(f"Testing lexical retrieval for: '{test_query}'")
    hits = retrieve_kb(test_query, top_k=2)
    for h in hits:
        print(f"[{h['id']}] {h['title']} (Score: {h['score']}, Threshold: {h['passes_threshold']})")
